libraries/node-core-library/src/LockFile.ts

```diff
@@ -0,0 +1,130 @@
+    const items: string[] = psSplit[1].split(/\s+/);
+    return items[process.platform === 'darwin' ? 5 : 4];
+  }
+  return undefined;
+}
+
+/**
+ * A helper utility for working with file-based locks.
+ * This class should only be used for locking resources across processes,
+ * but should not be used for attempting to lock a resource in the same process.
+ * @public
+ */
+export class LockFile {
+  /**
+   * Attempts to create a lockfile with the given filePath.
+   * If successful, returns a LockFile instance.
+   * If unable to get a lock, returns undefined.
+   * @param filePath - the filePath of the lockfile.
+   */
+  public static tryAcquire(filePath: string): LockFile | undefined {
+    let dirtyWhenAcquired: boolean = false;
+
+    if (fsx.existsSync(filePath)) {
+
+      dirtyWhenAcquired = true;
+
+      if (process.platform === 'win32') {
+        // If the lockfile is held by an process with an exclusive lock, then removing it will
+        // silently fail. OpenSync() below will then fail and we will be unable to create a lock.
+
+        // Otherwise, the lockfile is sitting on disk, but nothing is holding it, implying that
+        // the last process to hold it died.
+      } else if (process.platform === 'darwin' || process.platform === 'linux') {
+        // we need to compare the process ID and start time in the lockfile
+        // to the result of the ps command
+
+        const contents: string = fsx.readFileSync(filePath).toString();
+        const [pid, startTime] = contents.split(';');
+
+        if (!!pid && !!startTime) {
+          const oldStartTime: string | undefined = getProcessStartTime(pid);
+          // the process that created this lock is still running!
+          if (oldStartTime === startTime) {
+            return undefined;
+          }
+        }
+      } else {
+        throw new Error(`Unsupported operating system: ${process.platform}`);
+      }
+
+      // If we get here it is safe to unlink the lockfile
+      fsx.unlinkSync(filePath);
+    }
+
+    let fileDescriptor: number;
+    try {
+      // Attempt to open an exclusive lockfile
+      fileDescriptor = fsx.openSync(filePath, 'wx');
+    } catch (error) {
+      // we tried to delete the lock, but something else is holding it,
+      // (probably an active process), therefore we are unable to create a lock
+      return undefined;
+    }
+
+    if (process.platform === 'darwin' || process.platform === 'linux') {
+      fsx.writeSync(fileDescriptor, `${process.pid};${getProcessStartTime(process.pid.toString())}`);
+    }
+
+    return new LockFile(fileDescriptor, filePath, dirtyWhenAcquired);
+  }
+
+  /**
+   * Unlocks a file and removes it from disk.
+   * This can only be called once.
+   */
+  public release(): void {
+    if (this.isReleased) {
+      throw new Error(`The lock for file "${path.basename(this._filePath)}" has already been released.`);
+    }
+
+    fsx.closeSync(this._fileDescriptor!);
+    fsx.removeSync(this._filePath);
```

Comment:
>fsx.removeSync(this._filePath); [](start = 4, length = 31)

If this throws an exception, what state are we in?

Reply:
The only case where this would throw would be an EPERM error.. ?

---
In reply to: [165796437](https://github.com/Microsoft/web-build-tools/pull/517#discussion_r165796437) [](ancestors = 165796437)